doc/SONiC_VRRP_HLD.md

```diff
@@ -0,0 +1,582 @@
+degradation.
+Though VRRP standard protocol present in RFC 3768 is complete in itself, there are few
+limitations/drawbacks of the protocol:
+
+1. The point of failure which VRRP safeguards against is the interface on which VRRP instance is
+  present and the router as a whole. Even if all of the connectivity of master to the external network
+  fails, VRRP will still not trigger the failover to the backup gateway. Uplink interface tracking feature has been introduced to overcome this limitation
+
+2. Only master router sends the advertisements, and backups just listen to them. Since backups do
+  not send any periodic message, it is difficult to ascertain that the backups are still active, and the
+  healthiness of the backup routers can not be known until the failover happens.
+
+3. VRRP owner - if a VRRP instance's virtual IP address is same as the real interface IP address. VRRP owner has special privileges (priority is 255) and can pre-empt any other router acting as master. This usually causes unwanted temporary network disruption after the non-owner master had stabilized
+  after the failover from owner.
+
+4. A non-owner master can not accept packets destined to the virtual IP address. Only the owner can
+  accept and respond to such packets.
+
+5. Two virtual routers with same VRIDs but on different IP subnets can not co-exist on the same
+  VLAN because of the virtual MAC address clash.
+
+  VRRP implementation in SONIC completely addresses limitation #1 by introducing uplink interface tracking feature.
+
+### 3.2.1    Pre-emption
+
+A VRRP Master node can be pre-empted by another VRRP node that has higher priority. Pre-emption is enabled by default. CLI is available for user to disable pre-emption on per VRRP instance. Mastership switchover causes unnecessary temporary network disruption.
+
+### 3.2.2 Virtual Router Owner
+
+VRRP instance whose virtual IP address (VIP) is same as real interface address is called the owner of virtual router and has the highest priority (255). VRRP owner is supported in SONIC's VRRP implementation. VRRP owner node will pre-empt the active master even when pre-emption is disabled for that particular VRRP instance.
+
+### 3.2.3 Virtual MAC Address
+
+Following virtual MAC address is used by the protocol (as per RFC 5798)
+
+IPv4 case: **00-00-5e-00-01-{vrid}**
+
+where, vrid is user configured 1-byte virtual router identifier. VRID has interface scope; that is, VRID has to be unique among the VRRP instances on an interface. However, same VRID can be used for two  or more virtual router instances across different interfaces.
+
+### 3.2.4    VRRP Advertisement Frame
+
+VRRP control packets have IP protocol type as 112 (reserved for VRRP), and are sent to VRRP multicast address 224.0.0.18. Source MAC in VRRP control packets is virtual MAC address and source IP is interface IP.                                                               
```

Comment:
we need sai trap id for this. do we have it defined in the SAI, if yes, then add this as a sai requirement.

Reply:
The already defined SAI trap IDs viz., SAI_HOSTIF_TRAP_TYPE_VRRP & SAI_HOSTIF_TRAP_TYPE_VRRPV6 will be used to trap the VRRP packets.